org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/classfmt/ClassFileReader.java

```diff
@@ -156,19 +159,24 @@ public static ClassFileReader readFromModule(
 		throws ClassFormatException, java.io.IOException {
 		return JRTUtil.getClassfile(jrt, filename, moduleName, moduleNameFilter);
 }
-public static ClassFileReader read(
-	java.util.zip.ZipFile zip,
-	String filename,
-	boolean fullyInitialize)
-	throws ClassFormatException, java.io.IOException {
+
+public static ClassFileReader read(java.util.zip.ZipFile zip, String filename, boolean fullyInitialize)
+		throws ClassFormatException, java.io.IOException {
 	java.util.zip.ZipEntry ze = zip.getEntry(filename);
-	if (ze == null)
+	if (ze == null) {
 		return null;
-	ClassFileReader classFileReader = Util.getZipEntryClassFile(zip.getName(), filename);
-	if (fullyInitialize) {
-		classFileReader.initialize();
 	}
-	return classFileReader;
+	try (BufferedInputStream stream = new BufferedInputStream(zip.getInputStream(ze))) {
+		Path zipfile = Paths.get(zip.getName());
+		URI uri1 = URI.create("jar:file://" + zipfile.toUri().getRawPath() + "!/" + filename); //$NON-NLS-1$ //$NON-NLS-2$
```

Comment:
inline `uri1`

Reply:
Also we might want to move toJarUri to this class.

Reply:
There might even be room for improvements.
A ZipEntry knows its uncompressed size in bytes. We could allocate the byte array with the correct size and call `read(byte[], offset, len)` in a loop until it's fully populated. This should spare us all the intermediate allocations.

Reply:
You should report that to JDK. Probably InflaterInputStream.readAllBytes() could return InflaterInputStream.buf for example

Reply:
> Also we might want to move toJarUri to this class.

its not used anywhere else. there is still a org.eclipse.core.runtime.URIUtil.toJarURI(URI, IPath)

Reply:
> inline `uri1`

done

Reply:
> You should report that to JDK. Probably InflaterInputStream.readAllBytes() could return InflaterInputStream.buf for example

I don't understand what you're suggesting.

ZipEntry offers `getSize` which returns the uncompressed size

```
byte[] bytes = new byte[ze.getSize()];
stream.readNBytes(bytes, 0, bytes.length);
```
*Should* be the fastest way to read the bytes in this particular case.

Reply:
stream.readAllBytes() should be the fastest way to read all bytes. if it is not the stream implementation should be improved. JDK already did that for Files, with massive performance boost

Reply:
stream.readAllBytes() allocates a number of intermediate byte arrays that will be concatenated in the end to a new one. I don't see how that can be faster than allocating a single array and reading into that one directly.

Reply:
Note that it is not the stream implementation's "fault" or responsibility. The stream can/should not make any assumptions about the offset within the stream when readAllBytes is called. Here, we want to read all bytes from start to end. Therefore a targeted optimization is possible, which would not be possible for the inflater stream without breaking encapsulation or maintaining additional state.

Reply:
A stream provided by ZipFile does know its size itself. see for example https://bugs.openjdk.org/browse/JDK-8264777 how good JDK solved such reports in the past.

Reply:
Ok, fair enough. I'd argue optimize locally for the time being until a general fix is available, but I understand your point.